### PR TITLE
Added Salesforce provider

### PR DESCRIPTION
**Pull request by amattie**

### CODE REVIEW

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException("An error occurred when retrieving the user from the Salesforce identity service.");
+            }
+
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+            Logger.LogTrace(payload.ToString());
+            identity.AddOptionalClaim(ClaimTypes.NameIdentifier, SalesforceAuthenticationHelper.GetUserIdentifier(payload), Options.ClaimsIssuer)
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs]: Omagad, this list is insane :smile: 
Adding all these claims will result in extremely large cookies, that may excess the browsers size limits.

Please consider reducing the list to 5/6 claims.

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHelper.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        /// <summary>
+        /// Gets the Salesforce ID corresponding to the authenticated user.
+        /// </summary>
+        public static string GetUserIdentifier([NotNull] JObject user) => user.Value<string>("user_id");
+
+        /// <summary>
+        /// Gets the user's name
+        /// </summary>
+        public static string GetUserName([NotNull] JObject user) => user.Value<string>("user_name");
+    }
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHelper.cs]: You'll need to add more helpers for the "claims" you want to support.

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationOptions.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+using Microsoft.AspNetCore.Http;
+
+namespace AspNet.Security.OAuth.Salesforce
+{
+    /// <summary>
+    /// Defines a set of options used by <see cref="SalesforceAuthenticationHandler"/>.
+    /// </summary>
+    public class SalesforceAuthenticationOptions : OAuthOptions
+    {
+        public SalesforceAuthenticationOptions OverrideHostInEndpoints(string newHost)
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationOptions.cs]: Curious: what's the concrete use case for this method?

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens)
+        {
+            var identityServiceUrl = tokens.Response.Value<string>("id");
+            var salesforceInstanceUrl = tokens.Response.Value<string>("instance_url");
+
+            var address = QueryHelpers.AddQueryString(identityServiceUrl, "token", tokens.AccessToken);
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs]: Is it really necessary to flow the access token in both the query string and the authorization header?

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        public SalesforceAuthenticationHandler([NotNull] HttpClient client)
+            : base(client)
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens)
+        {
+            var identityServiceUrl = tokens.Response.Value<string>("id");
+            var salesforceInstanceUrl = tokens.Response.Value<string>("instance_url");
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs]: So the "userinfo" endpoint is not static (which explains why it doesn't have a default value in the options) and must be resolved at runtime?

**Comment by PinpointTownes** [comment]
Thanks for your contribution! :clap: 

Minor a few remarks/questions, it looks fine.

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationOptions.cs (hunk 3 above)
Comment (amattie) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationOptions.cs]: It's pretty common in Salesforce land to think of and use test.salesforce.com in dev / test / stage environments and login.salesforce.com in production environments. At the same time, it seems unnecessary to have to change the endpoints to the full https://test.salesforce.com/services/oauth2/authorize and https://test.salesforce.com/services/oauth2/token for those non-prod environments, so this method makes it easier to simply override the hosts without having to look up and set the full endpoints when first getting this going.

That said, your lib, and so I'm happy to remove if you'd prefer.

Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationOptions.cs]: No, it's fine. But I wonder if an `Environment` enum property wouldn't be clearer. We already use a similar pattern in the BattleNet provider for regions:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/src/AspNet.Security.OAuth.BattleNet/BattleNetAuthenticationOptions.cs#L30

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        public SalesforceAuthenticationHandler([NotNull] HttpClient client)
+            : base(client)
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens)
+        {
+            var identityServiceUrl = tokens.Response.Value<string>("id");
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs]: Curious: is there official documentation about this `id` property? It seems to be an odd name for an URL, no?

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            }
+
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+
+            identity.AddOptionalClaim(ClaimTypes.NameIdentifier, SalesforceAuthenticationHelper.GetUserIdentifier(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim(ClaimTypes.Name, SalesforceAuthenticationHelper.GetUserName(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim("urn:salesforce:email", SalesforceAuthenticationHelper.GetEmail(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim("urn:salesforce:thumbnail_photo", SalesforceAuthenticationHelper.GetThumbnailPhoto(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim("urn:salesforce:utc_offset", SalesforceAuthenticationHelper.GetUtcOffset(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim("urn:salesforce:rest_url", SalesforceAuthenticationHelper.GetRestUrl(payload), Options.ClaimsIssuer);
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs]: Question: what kind of document is returned by this mysterious REST URL?

File: src/AspNet.Security.OAuth.Salesforce/AspNet.Security.OAuth.Salesforce.xproj [hunk 8]
```diff
@@ -0,0 +1,18 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">14.0</VisualStudioVersion>
+    <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
+  </PropertyGroup>
+  <Import Project="$(VSToolsPath)\DNX\Microsoft.DNX.Props" Condition="'$(VSToolsPath)' != ''" />
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/AspNet.Security.OAuth.Salesforce.xproj]: Note for me: this project file must be normalized to match the MSBuild files used by the new tooling.

File: src/AspNet.Security.OAuth.Salesforce/project.json [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  "dependencies": {
+    "AspNet.Security.OAuth.Extensions": { "type": "build", "version": "1.0.0-*" },
+    "JetBrains.Annotations": { "type": "build", "version": "10.1.2-eap" },
+    "Microsoft.AspNetCore.Authentication.OAuth": "1.0.0-*"
+  },
+
+  "frameworks": {
+    "net451": {
+      "frameworkAssemblies": {
+        "System.Runtime": { "type": "build" }
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/project.json]: Note for me: this reference assembly can be removed, as it's no longer required with the new VS tooling.

File: src/AspNet.Security.OAuth.Salesforce/project.json [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+  "buildOptions": {
+    "warningsAsErrors": true,
+    "nowarn": [ "CS1591" ],
+    "xmlDoc": true
+  },
+
+  "dependencies": {
+    "AspNet.Security.OAuth.Extensions": { "type": "build", "version": "1.0.0-*" },
+    "JetBrains.Annotations": { "type": "build", "version": "10.1.2-eap" },
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/project.json]: Note for me: this package must be upgraded to 10.1.4 to match the changes committed yesterday.

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs (hunk 6 above)
Comment (amattie) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs]: Yeah, agreed, it's pretty silly. It's documented on this page: https://developer.salesforce.com/page/Digging_Deeper_into_OAuth_2.0_on_Force.com. CTRL+F for "A URL, representing the authenticated user". First match halfway down page is a description of the object that's returned in the initial OAuth flow.

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs (hunk 7 above)
Comment (amattie) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs]: They return a giant JSON object. I just had to pick through what I'd guess others would want and what I've used most in the past. Documented here: https://developer.salesforce.com/page/Digging_Deeper_into_OAuth_2.0_on_Force.com#The_Force.com_Identity_Service

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationOptions.cs (hunk 3 above)
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationOptions.cs]: No particular remark on my suggestion? :smile:

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs (hunk 6 above)
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs]: Great, thanks! :clap: 

Note for me: add this link as a comment before merging.

File: src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs (hunk 7 above)
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs]: Okay, thanks for the precision.

**Comment by PinpointTownes** [comment]
Sorry for the late feedback :smile: 

Minor my remark/question about `OverrideHostInEndpoints`, it looks fine (I'll merge it soon).

**Comment by amattie** [comment]
Hang on. I have your environment suggestion implemented, but I haven't been
able to test yet due to the RC1 -> 2 switch. I'll test + push later today.

**Comment by amattie** [comment]
Ok, I changed the way environments are handled to your suggestion. Everything should be good now.

**Comment by PinpointTownes** [comment]
Looks fine, I'll merge it tomorrow :clap:

**Comment by PinpointTownes** [comment]
Merged (sorry for the delay :sweat_smile:): https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/commit/14618f683087936a784b1191215f857583ae880f

Thanks for your contribution!

**Pull request closed without merging**